subprojects/core/src/main/java/org/gradle/caching/internal/controller/impl/RootBuildCacheControllerRef.java

```diff
@@ -17,63 +17,179 @@
+    public LifecycleAwareBuildCacheController createForNonRootBuild(Path identityPath, BuildCacheControllerFactory buildCacheControllerFactory, InstanceGenerator instanceGenerator) {
+        return rootController.createChild(identityPath, buildCacheControllerFactory, instanceGenerator);
     }
 
-    public BuildCacheController getForNonRootBuild() {
-        if (!isSet()) {
-            throw new IllegalStateException("Root build cache controller not yet assigned");
+    private static abstract class DelegatingBuildCacheController implements LifecycleAwareBuildCacheController {
+        @Override
+        public boolean isEnabled() {
+            return getDelegate().isEnabled();
         }
 
-        return buildCacheController;
-    }
+        @Override
+        public Optional<BuildCacheLoadResult> load(BuildCacheKey cacheKey, CacheableEntity cacheableEntity) {
+            return getDelegate().load(cacheKey, cacheableEntity);
+        }
+
+        @Override
+        public void store(BuildCacheKey cacheKey, CacheableEntity entity, Map<String, FileSystemSnapshot> snapshots, Duration executionTime) {
+            getDelegate().store(cacheKey, entity, snapshots, executionTime);
+        }
+
+        @Override
+        public void close() {
+            resetState();
+        }
+
+        protected static void createDelegate(BuildCacheConfigurationInternal configuration, AtomicReference<BuildCacheController> delegate, BuildCacheControllerFactory buildCacheControllerFactory, Path identityPath, InstanceGenerator instanceGenerator) {
+            BuildCacheController controller = buildCacheControllerFactory.createController(identityPath, configuration, instanceGenerator);
+            if (!delegate.compareAndSet(NoOpBuildCacheController.INSTANCE, controller)) {
+                try {
+                    controller.close();
+                } catch (IOException e) {
+                    throw UncheckedException.throwAsUncheckedException(e);
+                }
+                throw new IllegalStateException("Build cache controller already set");
+            }
+        }
 
-    public boolean isSet() {
-        return buildCacheController != null;
+        protected static void discardDelegate(AtomicReference<BuildCacheController> delegate) {
+            BuildCacheController controller = delegate.getAndSet(NoOpBuildCacheController.INSTANCE);
+            try {
+                controller.close();
+            } catch (IOException e) {
+                throw UncheckedException.throwAsUncheckedException(e);
+            }
+        }
+
+        protected abstract BuildCacheController getDelegate();
     }
 
-    private static class CloseShieldBuildCacheController implements BuildCacheController {
-        private final BuildCacheController delegate;
+    private static class RootBuildCacheController extends DelegatingBuildCacheController {
+        private final AtomicReference<BuildCacheController> delegate = new AtomicReference<>();
```

Comment:
Holds an immutable BCC that represents the current state of the root build's cache configuration.